pfn_notify callback added for clBuildPorgram, related test added and -Werror test revised

ardacoskunses opened the pull request:
pfn_notify callback added for clBuildProgram, and related test added to clBuildProgram tests.

OCL1.2 spec says:
"If pfn_notify is not NULL, clBuildProgram does not need to wait for the build to complete and can return immediately once the build operation can begin."
This feature still missing.

Also -Werror test revised to test full functionality: warning in the kernel build turned to error and checked in the test.

pjaaskel commented on lib/CL/clBuildProgram.c:
```diff
@@ -204,6 +205,21 @@ CL_API_SUFFIX__VERSION_1_0
 
   POCL_LOCK_OBJ(program);
 
+  if(pfn_notify)
```
Please add white space after if. Check https://www.gnu.org/prep/standards/html_node/Formatting.html

pjaaskel commented on lib/CL/clBuildProgram.c:
```diff
@@ -495,6 +511,11 @@ CL_API_SUFFIX__VERSION_1_0
   POCL_UNLOCK_OBJ(program);
 
+  /* callback freed in ProgramRelease */
+  if(program->buildprogram_callbacks)
```
Ditto.

pjaaskel commented on lib/CL/clBuildProgram.c:
```diff
@@ -520,6 +541,12 @@ CL_API_SUFFIX__VERSION_1_0
    * cause a double free. */
 
 ERROR:
+  if(program->buildprogram_callbacks)
```
Here also.

pjaaskel commented on lib/CL/clBuildProgram.c:
```diff
@@ -204,6 +205,21 @@ CL_API_SUFFIX__VERSION_1_0
+      callback = (build_program_callback_t*) malloc (sizeof (build_program_callback_t));
+      if (callback == NULL)
+      {
+        POCL_UNLOCK_OBJ(program);
```
Also white space is needed before the argument list (in the GNU C style).

pjaaskel commented on lib/CL/clBuildProgram.c:
```diff
@@ -520,6 +541,12 @@ CL_API_SUFFIX__VERSION_1_0
+    {
+      program->buildprogram_callbacks->callback_function (program,
+                         program->buildprogram_callbacks->user_data);
+      POCL_MEM_FREE(program->buildprogram_callbacks);
```
Ditto.

pjaaskel commented on lib/CL/clReleaseProgram.c:
```diff
@@ -59,6 +59,9 @@ POname(clReleaseProgram)(cl_program program) CL_API_SUFFIX__VERSION_1_0
           k->program = NULL;
         }
 
+      if(program->buildprogram_callbacks)
```
Ditto (twice).

pjaaskel commented on lib/CL/clReleaseProgram.c:
```diff
@@ -59,6 +59,9 @@ POname(clReleaseProgram)(cl_program program) CL_API_SUFFIX__VERSION_1_0
         }
 
+      if(program->buildprogram_callbacks)
+        POCL_MEM_FREE(program->buildprogram_callbacks);
```
Isn't this a linked list?

pjaaskel commented on lib/CL/clBuildProgram.c:
```diff
@@ -495,6 +511,11 @@ CL_API_SUFFIX__VERSION_1_0
 
+  /* callback freed in ProgramRelease */
+  if(program->buildprogram_callbacks)
+      program->buildprogram_callbacks->callback_function (program,
```
Shouldn't it traverse the linked list?

ardacoskunses commented:
Agreed all the comments and updated the PR accordingly.

Linked list implementation is not necessary for clBuildProgram pfn_notify so I changed to single callback.
**Note on this:** all implemented callback functions in pocl are linkedlist (clSetMemObjectDestructorCallback, and clEventCallback) OCL1.1 spec explains their usage scenarios as multiple callbacks, so it makes sense using linked list for them. 
My understanding from spec clBuildProgram callback does **not** have to be linked list. 
Please let me know if otherwise.

pjaaskel commented on include/pocl.h:
```diff
@@ -58,6 +58,14 @@ struct _mem_destructor_callback
 };
 
+typedef struct _build_program_callback build_program_callback_t;
+/* represents a build program callback */
```
This comment is useless as it basically repeats the struct name.
